File: day2/puzzle2.py
""" Task: Find the two rows that differ by one character and print the common characters
"""
from itertools import combinations

def search_common(lines):
    for line1, line2 in combinations(lines, 2):
        numdiff = sum((c1 != c2 for c1, c2 in zip(line1, line2)))
        if numdiff == 1:
            return "".join((c1 for c1, c2 in zip(line1, line2) if c1 == c2))

def main():
    with open("input.txt") as f:
        lines = f.readlines()
    
    print(search_common(lines))

if __name__ == "__main__":
    main()

File: day2/puzzle2.jl
""" Task: Find the two rows that differ by one character and print the common characters
"""

function searchcommon(lines)

    for idx1 in 1:length(lines)-1, idx2 in idx1:length(lines)
        line1 = collect(lines[idx1])
        line2 = collect(lines[idx2])
        
        # Difference the arrays of Char
        diff = line1 .- line2 .!= 0
        if sum(line1 .- line2 .!= 0) == 1
            return join(line1[.~diff])
        end
    end
    return "failed"      
end


function main()
    lines = readlines(open("input.txt", "r"))
    println(searchcommon(lines))

end

main()
